### DISCUSSION TITLE
Mimicking three and examples folder

### DISCUSSION THREAD

**Discussion by swingingtom** [felixmariotto/three-mesh-ui#116] [Ideas]
Hey community,

with such features as #1 Text Selection , #114 Best Fit, #113 computeUV2 , wouldn't be a structure like an examples folder be interesting?

New features like effects, controls, behaviour could be out of `ThreeMeshUI` global object, and loaded only we project requires it.

This could reduce the bundle size, and keeps core from "not so frequent case" code complexity.

**Comment by felixmariotto** [comment]
Hi @swingingtom,
Yes that's what [this folder](https://github.com/felixmariotto/three-mesh-ui/tree/master/examples/utils) is about, but maybe we could rename it and create subfolders ?

**Comment by swingingtom** [comment]
As I can interpret it, right now the current state of example folder contains different purposes : 

- Feature demonstrations ( Kerning, anti aliasing, big text )
- Code samples ( basic setup, tutorial step )

And something like #121 or https://user-images.githubusercontent.com/2693840/150170272-742e07e4-6289-45a9-bb04-299583b3303b.png  which is really useful when developing or debugging a feature doesn't really fit in any of the previous cases. It might even be unproductive to publish them to users. 

> I'm mostly worried that we will duplicate most of what is in examples/assets and examples/utils... If we create a space for work-in-progress experiments I think it should still be in the example folder. Maybe an example/WIP folder ?

I think you are right about your concern. 
When testing or developing, we should cover most possible cases, we may also need custom or external tools such as gui.
Whereas, samples or examples aiming for users, should have the minimal amount of code to use that single feature / sample.
Sometimes they both conflicts.

Maybe just a flag on example to tell published or not might do the trick?

Maybe using a gui library and discarding it inside webpack build command (`env.NODE_ENV=prod`)?

**Comment by swingingtom** [comment]
@felixmariotto I started the branch `refactor-examples-folder`  which acts as a POC of what direction examples folder could take. 

## Uniformized imports
It tries to uniformize code written witin examples directory with what users code would have to be : 
```javascript
// unstandardized imports
import ThreeMeshUI from '../src/three-mesh-ui.js';

//standardized imports
import ThreeMeshUI from 'three-mesh-ui';
```
The same is done with pieces of code located within the examples directory
```javascript
// unstandardized imports
import VRControl from './utils/VRControl.js';

//standardized imports
import ThreeMeshUI from 'three-mesh-ui/examples/controls/VRControl.js';
```

The suggestion in this PR is to use webpack aliases to do so
```javascript
//webpack.prod.config.js
[...]
 resolve: {
            alias: {
                "three-mesh-ui/examples": path.resolve(__dirname, '../examples/'),
                "three-mesh-ui": path.resolve(__dirname, '../src/three-mesh-ui.js'),
            },
        },
```

## Categorizing examples
Using webpack.prod.config , we could categorize any root example js files into different categories : 
* tutorial (tut_) : Small pieces of code, in order for users to easily get started with three-mesh-ui just by reading source.
* examples (ex_): More complex pieces of code showing what can be built with and on-top of three-mesh-ui
* features (feat_) : Relatively complex pieces of code ( ideally interactive ) to demonstrate features and their available configurations ( best_fit and flex_layout(draft) show some possibility )
* workbenches (dev_) : Not published while webpack in production, this could be developers and tests and validation area. 

![image](https://user-images.githubusercontent.com/17314210/150770484-3b54aa05-c509-41e1-9db9-7521e7f307aa.png)


## Built-in examples components
Some examples create locally components that are really handy, and could be available to build more examples without having to copy/paste. 

Users could rely on those to use them as-is, or rely on those component code to have an examples on how components can be made using three-mesh-ui core components combinaisons.

Developers could rely on them to provide mode interactive feature demonstrations or workbenches.

Ui/Buttons. They comes from an example, but could be used in more than one examples. A 'ui' folder with more built-in ui components examples could be great.

Same goes for VRControl and Raycaster. Raycaster is used a two locations with different implementations (Buttons & Keyboard)

**Comment by swingingtom** [comment]
## Some toughts on UI/Interactive components
### JS / JSM
Currently all suggested components are jsm only. Not sure it worth to also support js, but Im clearly biaised.

### Keyboard
`ThreeMeshUI.Keyboard` might move to examples/ui. This is a great example of what can be built on-top of three-mesh-ui. Users and developers could use it as-is. This might slightly lighten three-mesh-ui build without removing its availability (except for js).

### States vs Styles for interactive components ( cumulative states )
If you look at the "Toggle Buttons" in the "interactive_components" page, it suffers from what I would call "100% relying on states". While clicking a button, it doesnt get green until we roll over it. Because 'hovered' and 'selected' should be cumulative. We could build ui components with "styles" properties instead of "states", and internally in ui components, redefining "states" according to cumulative states.

If the component is selected => lets redefined its hovered state with : hovered style overwritten by any selected style properties.
If the component is not more selected => lets redefined its hovered state with : hovered style.

### States triggering ( naming convention )
Actions requested by raycaster could be present tense. Ui component would compute its state.

Raycaster 'select'  => Ui component if can be selected => 'selected'
Raycaster 'hover' => Ui component if can be hovered => 'hovered'

### Event triggering
Unfortunately I based the branch components upon the raycaster of button, and raycaster of keyboard looked a bit more of what I would had need. 

But internally, a component could use its x latest states to know what state to trigger.

Raycaster 'select' => Ui component if latest states are ['idle','hovered'] => click
Raycaster 'select' => Ui component if latest state are ['hovered', 'idle' ] => not click

### Next
Those draft component weren't that difficult to built, but their current state are far from being completed. Having more discussed basis would easy their refactor. If you agree with.

#### Side note
Why do we have contentAlign "left,top,bottom,right" instead of "start,end"? I got a lot of wrong alignment issues.
